snuba/request/validation.py

```diff
@@ -215,3 +179,83 @@ def build_request(
 
         timer.mark("validate_schema")
         return request
+
+
+def _get_referrer(request_parts: RequestParts, referrer: str) -> str:
+    tenant_referrer = request_parts.attribution_info["tenant_ids"].get("referrer")
+    if tenant_referrer != referrer:
+        metrics.increment(
+            "referrer_mismatch",
+            tags={
+                "tenant_referrer": tenant_referrer or "none",
+                "request_referrer": referrer,
+            },
+        )
+    # Handle an edge case where the legacy endpoint is used.
+    return tenant_referrer or referrer
+
+
+def _get_settings_object(
+    settings_class: Type[HTTPQuerySettings] | Type[SubscriptionQuerySettings],
+    request_parts: RequestParts,
+    referrer: str,
+) -> HTTPQuerySettings | SubscriptionQuerySettings:
+    if settings_class == HTTPQuerySettings:
+        query_settings: MutableMapping[str, bool | str] = {
+            **request_parts.query_settings,
+            "consistent": _consistent_override(
+                request_parts.query_settings.get("consistent", False), referrer
+            ),
+        }
+        # TODO: referrer probably doesn't need to be passed in, it should be from the body
+        query_settings["referrer"] = referrer
+        # the parameters accept either `str` or `bool` but we pass in `str | bool`
+        return settings_class(**query_settings)  # type: ignore
+    elif settings_class == SubscriptionQuerySettings:
+        return settings_class(
+            consistent=_consistent_override(True, referrer),
+        )
+    return None  # type: ignore
+
+
+def _get_project_id(query: Query | CompositeQuery[Entity]) -> int | None:
+    project_ids = get_object_ids_in_query_ast(query, "project_id")
+    if project_ids is not None and len(project_ids) == 1:
+        return project_ids.pop()
+    return None
+
+
+def _get_attribution_info(
+    request_parts: RequestParts, referrer: str, query_project_id: int | None
+) -> AttributionInfo:
+    return AttributionInfo(
+        **update_attribution_info(request_parts, referrer, query_project_id)
+    )
+
+
+def _build_request(
```

Comment:
why did this function need to be pulled out of `build_request`? Does it have different concerns? Is it going to be reused somewhere else?

Reply:
Yes it is reused, I pulled it out to prevent duplicate code. It is called in the except clause as well as in the normal body.

I saw build_request as doing:
1. from raw request, parse into query
2. from query build request. (this is _build_request)

Step 2 was already done in the normal case, build now it had to be done in case of exception as well. So I factored it out into a method and used the method in both cases.